pkg/gofer/origins/uniswap_v3.go

```diff
@@ -0,0 +1,187 @@
+//  Copyright (C) 2020 Maker Ecosystem Growth Holdings, INC.
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Affero General Public License as
+//  published by the Free Software Foundation, either version 3 of the
+//  License, or (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Affero General Public License for more details.
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package origins
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/makerdao/oracle-suite/internal/query"
+)
+
+const uniswapV3URL = "https://api.thegraph.com/subgraphs/name/ianlapham/uniswap-v3-alt"
+
+type uniswapV3Response struct {
+	Data struct {
+		Pools []uniswapV3PairResponse
+	}
+}
+
+type uniswapV3TokenResponse struct {
+	Symbol string `json:"symbol"`
+}
+
+type uniswapV3PairResponse struct {
+	ID      string                 `json:"id"`
+	Price0  stringAsFloat64        `json:"token0Price"`
+	Price1  stringAsFloat64        `json:"token1Price"`
+	Volume0 stringAsFloat64        `json:"volumeToken0"`
+	Volume1 stringAsFloat64        `json:"volumeToken1"`
+	Token0  uniswapV3TokenResponse `json:"token0"`
+	Token1  uniswapV3TokenResponse `json:"token1"`
+}
+
+type UniswapV3 struct {
+	Pool query.WorkerPool
+}
+
+func (u *UniswapV3) pairsToContractAddress(pair Pair) string {
+	// We're checking for reverse pairs because the same contract is used to
+	// trade in both directions.
+	match := func(a, b Pair) bool {
+		if a.Quote == b.Quote && a.Base == b.Base {
+			return true
+		}
+
+		if a.Quote == b.Base && a.Base == b.Quote {
+			return true
+		}
+
+		return false
+	}
+
+	p := Pair{Base: u.renameSymbol(pair.Base), Quote: u.renameSymbol(pair.Quote)}
+
+	switch {
+	case match(p, Pair{Base: "COMP", Quote: "WETH"}):
+		return "0xea4ba4ce14fdd287f380b55419b1c5b6c3f22ab6"
+	case match(p, Pair{Base: "CRV", Quote: "WETH"}):
+		return "0x04916039b1f59d9745bf6e0a21f191d1e0a84287"
+	case match(p, Pair{Base: "WETH", Quote: "USDC"}):
+		return "0x8ad599c3a0ff1de082011efddc58f1908eb6e6d8"
+	case match(p, Pair{Base: "WETH", Quote: "KNC"}):
+		return "0x76838fd2f22bdc1d3e96069971e65653173edb2a"
+	case match(p, Pair{Base: "LRC", Quote: "WETH"}):
+		return "0xe1d92f1de49caec73514f696fea2a7d5441498e5"
```

Comment:
Let's hold off on KNC and LRC for now, volume is neglible for these on uniswap v3.